File: event-sourcing/python/tests/account_builder.py
from __future__ import annotations

from datetime import datetime, timedelta, timezone

from event_sourcing import (
    Account,
    AccountEvent,
    AccountOpened,
    MoneyDeposited,
    MoneyWithdrawn,
    AccountClosed,
    Money,
)

T0 = datetime(2026, 1, 1, tzinfo=timezone.utc)


class AccountBuilder:
    def __init__(self) -> None:
        self._account_id = "ACC-001"
        self._owner_name = "Alice"
        self._opened_at = T0
        self._additional_events: list[AccountEvent] = []
        self._time_step = 0

    def with_account_id(self, account_id: str) -> AccountBuilder:
        self._account_id = account_id
        return self

    def with_owner_name(self, name: str) -> AccountBuilder:
        self._owner_name = name
        return self

    def opened_at(self, timestamp: datetime) -> AccountBuilder:
        self._opened_at = timestamp
        return self

    def with_deposit(
        self, amount: float | int, timestamp: datetime | None = None
    ) -> AccountBuilder:
        self._additional_events.append(
            MoneyDeposited(
                account_id=self._account_id,
                timestamp=timestamp or self._next_timestamp(),
                amount=Money(amount),
            )
        )
        return self

    def with_withdrawal(
        self, amount: float | int, timestamp: datetime | None = None
    ) -> AccountBuilder:
        self._additional_events.append(
            MoneyWithdrawn(
                account_id=self._account_id,
                timestamp=timestamp or self._next_timestamp(),
                amount=Money(amount),
            )
        )
        return self

    def closed(self, timestamp: datetime | None = None) -> AccountBuilder:
        self._additional_events.append(
            AccountClosed(
                account_id=self._account_id,
                timestamp=timestamp or self._next_timestamp(),
            )
        )
        return self

    def build(self) -> Account:
        events: list[AccountEvent] = [
            AccountOpened(
                account_id=self._account_id,
                timestamp=self._opened_at,
                owner_name=self._owner_name,
            ),
            *self._additional_events,
        ]
        return Account.rebuild(events)

    def _next_timestamp(self) -> datetime:
        self._time_step += 1
        return self._opened_at + timedelta(hours=self._time_step)

File: event-sourcing/typescript/tests/AccountBuilder.ts
import { Account } from '../src/Account.js';
import type { AccountEvent } from '../src/Events.js';
import { accountOpened, moneyDeposited, moneyWithdrawn, accountClosed } from '../src/Events.js';
import { Money } from '../src/Money.js';

export class AccountBuilder {
  private static readonly T0 = new Date(Date.UTC(2026, 0, 1));

  private accountId = 'ACC-001';
  private ownerName = 'Alice';
  private openedAt = AccountBuilder.T0;
  private additionalEvents: AccountEvent[] = [];
  private timeStep = 0;

  withAccountId(id: string): this { this.accountId = id; return this; }
  withOwnerName(name: string): this { this.ownerName = name; return this; }
  openedAtTime(timestamp: Date): this { this.openedAt = timestamp; return this; }

  withDeposit(amount: number, timestamp?: Date): this {
    this.additionalEvents.push(
      moneyDeposited(this.accountId, new Money(amount), timestamp ?? this.nextTimestamp()),
    );
    return this;
  }

  withWithdrawal(amount: number, timestamp?: Date): this {
    this.additionalEvents.push(
      moneyWithdrawn(this.accountId, new Money(amount), timestamp ?? this.nextTimestamp()),
    );
    return this;
  }

  closed(timestamp?: Date): this {
    this.additionalEvents.push(
      accountClosed(this.accountId, timestamp ?? this.nextTimestamp()),
    );
    return this;
  }

  build(): Account {
    const events: AccountEvent[] = [
      accountOpened(this.accountId, this.ownerName, this.openedAt),
      ...this.additionalEvents,
    ];
    return Account.rebuild(events);
  }

  private nextTimestamp(): Date {
    this.timeStep++;
    return new Date(this.openedAt.getTime() + this.timeStep * 3600_000);
  }
}
